### PR TITLE
Split waypoint and ztunnel flags

### PR DESCRIPTION
**Pull request by howardjohn**
The goal here is to make it possible to split the alpha and beta aspects of ambient, and allow a user to explicitly turn on the core (ztunnels) vs the less stable (waypoint).

This introduces a new flag for waypoints specifically.

The intent is immediately after this, we will merge a PR to make ambient controllers on by default (pending https://github.com/istio/istio/pull/49642)

### CODE REVIEW

**Review by bleggett — APPROVED** [review]
Approach SGTM, LGTM pending CI.

**Comment by costinm** [comment]
Not sure I get the purpose of the extra knob.

Feature flags exist to protect existing users in case there is a bug - in this case the extra controllers. If the controllers run anyways - what's the risk here ? If there is a problem caused by generating configs for waypoints - user can remove the waypoints to mitigate. 

Also - as we get closer to beta, do we still expect the controllers to be a risk ? I was thinking of the opposite change - just make the controller enabled by default and remove the extra burden ( still keep them for a release or 2 in case something goes wrong or users want to disable )

**Comment by howardjohn** [comment]
> Also - as we get closer to beta, do we still expect the controllers to be a risk ? I was thinking of the opposite change - just make the controller enabled by default and remove the extra burden ( still keep them for a release or 2 in case something goes wrong or users want to disable )

That is the intent here. To make it so we can make the core 'beta' and on by default, but turn off the non-beta parts.

File: pilot/pkg/features/experimental.go
Comment (hzxuzhonghu) [inline pilot/pkg/features/experimental.go:183]: This is to make ambient conflex to understand,  the ideal way is: we have a flag indicating whether to enable ambient or not. 
I could not understand it even reading this description.

Comment (howardjohn) [reply pilot/pkg/features/experimental.go:183]: The user experience for ambient does not change: --set profile=ambient.

the purpose of this change is to make it so we can default enable a subset if feature's.

The current plan is to promote ztunnel to beta in 1.22. Beta means we need the controller on by default. Waypoints are alpha so they are inappropriate to turn on by default.

**Comment by hanxiaop** [comment]
> That is the intent here. To make it so we can make the core 'beta' and on by default, but turn off the non-beta parts.

WDYT about having a flag like enable_ambient_alpha instead of one that is specific to waypoints?

File: pilot/pkg/features/experimental.go
Comment (hzxuzhonghu) [reply pilot/pkg/features/experimental.go:183]: waypoint is the most important component of ambient, without it how can we call beta

Comment (dhawton) [reply pilot/pkg/features/experimental.go:183]: I disagree it being the "most important". Important, yes, but not *the* most important. Not everyone needs/wants L7 processing.

**Review by dhawton — APPROVED** [review]
LGTM and I agree with this approach.

**Comment by bleggett** [comment]
> > That is the intent here. To make it so we can make the core 'beta' and on by default, but turn off the non-beta parts.
> 
> WDYT about having a flag like enable_ambient_alpha instead of one that is specific to waypoints?

That occurred to me as well, but I think I would rather have feature flags specifically, so that it is clear what they do.

Having a generic `ambient_alpha` flag does not really tell you what exactly it turns on and off, and makes it easy to hide lots of confusing functionality behind a single flag, which is something we probably should not do.

**Comment by hanxiaop** [comment]
> > > That is the intent here. To make it so we can make the core 'beta' and on by default, but turn off the non-beta parts.
> > 
> > 
> > WDYT about having a flag like enable_ambient_alpha instead of one that is specific to waypoints?
> 
> That occurred to me as well, but I think I would rather have feature flags specifically, so that it is clear what they do.
> 
> Having a generic `ambient_alpha` flag does not really tell you what exactly it turns on and off, and makes it easy to hide lots of confusing functionality behind a single flag, which is something we probably should not do.
> 
> The compatibility versions effectively are this anyway, but better documented/higher-level.

I was thinking it's similar to enabling alpha/experimental APIs, which doesn't have a worse usability. But regardless of this, my actual point of confusion is that the waypoint is not configured automatically upon installation; it's enabled based on the user's requirements. Why don't we just document the instability once L4 is beta and allow them to use it at their own risk, instead of having this specific flag for controllers?

File: pilot/pkg/features/experimental.go
Comment (hzxuzhonghu) [reply pilot/pkg/features/experimental.go:183]: Oh, no. This breaks my knowledge. From the begining, L7 is high lighted. https://istio.io/v1.15/blog/2022/introducing-ambient-mesh/

can users imagine a mesh does not support L7 features

Comment (dhawton) [reply pilot/pkg/features/experimental.go:183]: Right, highlighted.. but it's not *the most* important. It is one important component that makes up ambient, but it isn't more important than say the CNI plugin or ztunnel. Delaying L4/ztunnel moving to beta for L7/waypoint doesn't make sense, because they are different components. There are some users that don't necessary want/need the L7.

Comment (hzxuzhonghu) [reply pilot/pkg/features/experimental.go:183]: It is good to announce waypoint is alpha. And it equals ambient l4 beta  but not ambient beta

**Comment by costinm** [comment]
On Thu, Feb 29, 2024, 18:57 John Howard ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pilot/pkg/features/experimental.go
> <https://github.com/istio/istio/pull/49644#discussion_r1508424439>:
>
> > @@ -180,6 +180,17 @@ var (
>  		false,
>  		"If enabled, controllers required for ambient will run. This is required to run ambient mesh.").Get()
>
> +	EnableAmbientWaypoints = func() bool {
>
> The user experience for ambient does not change: --set profile=ambient.
>
> the purpose of this change is to make it so we can default enable a subset
> if feature's.
>
> The current plan is to promote ztunnel to beta in 1.22. Beta means we need
> the controller on by default. Waypoints are alpha so they are inappropriate
> to turn on by default.
>
Where does this rule come from ?

Not aware of any requirement that beta features are on by default, and we
have countless non-beta features available post install.

Unless this is coupled with the 'stable/experimental' separation - and ALL
the non stable stuff is turned off - I don't think Ambient should be
treated differently.

**Review by keithmattix — APPROVED** [review]
LGTM I think this is necessary with us planning to promote just L4

**Comment by howardjohn** [comment]
> Where does this rule come from ?

We do not _have_ to enable it by default, but it probably makes sense to. This is just the ability to correctly serve WDS really.

At the very least, if a user turns on a Beta feature they must get stability. Today, we have 1 flag to turn on Beta and experimental code, which is not safe.

> It is good to announce waypoint is alpha. And it equals ambient l4 beta but not ambient beta

The current messaging planned by Steering around this is that we are promoting "Ambient L4 to Beta"

File: pilot/pkg/features/experimental.go
Comment (zengyuxing007) [reply pilot/pkg/features/experimental.go:183]: > I disagree it being the "most important". Important, yes, but not _the_ most important. Not everyone needs/wants L7 processing.

that the needs of users in different locations may vary.
Many users in China maybe not need ztunnel, while waypoint L7 is essential.

Because most users in China do not have a need for FIPS. And some of the features of L4 can be adopted from Kuberntes CNI, such as NetworkPolicy, although it is not as good as AuthoriztionPolicy, but for many users, it is enough.

I'd like to propose a PR to see if it's possible to not enable most of ztunnel's features, but only enable DNS proxy, and use dnx proxy forward the traffic directly to waypoint.

**Comment by costinm** [comment]
On Wed, Mar 6, 2024 at 7:18 PM zengyuxing007 ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pilot/pkg/features/experimental.go
> <https://github.com/istio/istio/pull/49644#discussion_r1515434131>:
>
> > @@ -180,6 +180,17 @@ var (
>  		false,
>  		"If enabled, controllers required for ambient will run. This is required to run ambient mesh.").Get()
>
> +	EnableAmbientWaypoints = func() bool {
>
> I disagree it being the "most important". Important, yes, but not *the*
> most important. Not everyone needs/wants L7 processing.
>
> that the needs of users in different locations may vary.
> Many users in China maybe not need ztunnel, while waypoint L7 is essential.
>
> Because most users in China do not have a need for FIPS. And some of the
> features of L4 can be adopted from Kuberntes CNI, such as NetworkPolicy,
> although it is not as good as AuthoriztionPolicy, but for many users, it is
> enough.
>
It's not only China - even in the US very few users need FIPS, and even if
they did - Istio does not provide FIPS.

IMO  'ambient L4' effectively means 'a secure L4 transport' plus ability to
do some L4 redirection.
ztunnel is one possible implementation, and it happens to use mTLS and
HTTP/2 CONNECT (== hbone), but the
critical part is the CNI layer that handles the capture, and the XDS
extensions that control the redirection and
L4 policies.

The protocol and encryption parts are swappable - if Kubernetes CNI has
acceptable security, it should be possible
for a vendor to modify the code and remove mTLS. Or not use http/2.

But the controllers in Istiod that tell the L4 where to send the traffic
and the L4 authz policies are critical - and that's what this
API enables.



> I'd like to propose a PR to see if it's possible to not enable most of
> ztunnel's features, but only enable DNS proxy, and use dnx proxy forward
> the traffic directly to waypoint.
>
I have very big concerns about the DNS proxy ( and the security around it),
so may be fixable - but what you describe can also be done using private
DNS and custom
names - and I don't know if it needs any special PR, it is mostly possible
today ( with regular gateways ).

 Using DNS is an alternative to using L4 XDS controls, nice thing is that
we don't have to write any code, but it does require users to
make small config changes and think differently. We still want to have the
XDS controllers enabled by this PR.

John - my concern was more about using env variables or install-time
settings for users to 'opt in' to using Waypoints.
I think we have other alpha features that are opt-in by user creating a CRD
or setting an annotation - without having to
reinstall istio. Would it be hard to check if any Gateway has the class
'waypoint' to generate the required config ?

Enabling Istiod controllers for L4 seems pretty good.

**Comment by hzxuzhonghu** [comment]
The divergence focuses on whether a flag is a must.

If we donot have the new flag, it does not influence ambient L4 beta. If user want to try waypoint feature, need to create a `gateway`.

If this flag is there and false. User cannot try the L7 features even create a `gateway`. 

User previously have sued 1.20 or 1.21, they can continue on using waypoint if we donot have this flag. Otherwise, this will break them after upgrade to 1.22.

**Comment by zengyuxing007** [comment]
> On Wed, Mar 6, 2024 at 7:18 PM zengyuxing007 ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In pilot/pkg/features/experimental.go <[#49644 (comment)](https://github.com/istio/istio/pull/49644#discussion_r1515434131)>: > @@ -180,6 +180,17 @@ var ( false, "If enabled, controllers required for ambient will run. This is required to run ambient mesh.").Get() + EnableAmbientWaypoints = func() bool { I disagree it being the "most important". Important, yes, but not *the* most important. Not everyone needs/wants L7 processing. that the needs of users in different locations may vary. Many users in China maybe not need ztunnel, while waypoint L7 is essential. Because most users in China do not have a need for FIPS. And some of the features of L4 can be adopted from Kuberntes CNI, such as NetworkPolicy, although it is not as good as AuthoriztionPolicy, but for many users, it is enough.
> It's not only China - even in the US very few users need FIPS, and even if they did - Istio does not provide FIPS. IMO 'ambient L4' effectively means 'a secure L4 transport' plus ability to do some L4 redirection. ztunnel is one possible implementation, and it happens to use mTLS and HTTP/2 CONNECT (== hbone), but the critical part is the CNI layer that handles the capture, and the XDS extensions that control the redirection and L4 policies. The protocol and encryption parts are swappable - if Kubernetes CNI has acceptable security, it should be possible for a vendor to modify the code and remove mTLS. Or not use http/2. But the controllers in Istiod that tell the L4 where to send the traffic and the L4 authz policies are critical - and that's what this API enables.
> I'd like to propose a PR to see if it's possible to not enable most of ztunnel's features, but only enable DNS proxy, and use dnx proxy forward the traffic directly to waypoint.
> I have very big concerns about the DNS proxy ( and the security around it), so may be fixable - but what you describe can also be done using private DNS and custom names - and I don't know if it needs any special PR, it is mostly possible today ( with regular gateways ). Using DNS is an alternative to using L4 XDS controls, nice thing is that we don't have to write any code, but it does require users to make small config changes and think differently. We still want to have the XDS controllers enabled by this PR. John - my concern was more about using env variables or install-time settings for users to 'opt in' to using Waypoints. I think we have other alpha features that are opt-in by user creating a CRD or setting an annotation - without having to reinstall istio. Would it be hard to check if any Gateway has the class 'waypoint' to generate the required config ? Enabling Istiod controllers for L4 seems pretty good.
> […](#)
> Message ID: ***@***.***>

sorry,  I might think that FIPS and zero trust are pretty much the same, I mean there is no demand for zero trust in China.

ztunnel probably exists more because of the zero-trust requirement  because let the  traffic redirection to waypoint , which can be used in other ways, is not necessary to need ztunnel .
Some users which not  need for zero trust/mtls even use something like an ingress gateway to proxy east-west traffic.

I just think that waypoint is the feature that most users need.

**Comment by howardjohn** [comment]
@istio/wg-environments-maintainers ptal

**Review by zirain — APPROVED** [review]

**Review by hzxuzhonghu — APPROVED** [review]
Short term solution, in the long run after beta we only need one flag

**Pull request merged**
